### PR TITLE
feat(slice-4b): streaming q&a sse endpoint (pr4)

### PR DESCRIPTION
**Pull request by leehopper**
## Slice 4B PR4 — Streaming Q&A SSE endpoint (the integration gate)

Wires PR1 (`ICoachingLlm.StreamAsync`), PR2 (user-scoped `Conversation` persistence), and PR3 (intent classifier + `ComposeForConversationAsync`) behind a new **`POST /api/v1/conversation/messages`** Server-Sent-Events endpoint — the first interactive free-text chat path in the codebase. Spec: `docs/specs/19-spec-slice-4b-conversation-core/` (Unit 2). Decisions: DEC-073, DEC-085, DEC-084, DEC-059, DEC-060.

### What it does

Per message, `ConversationStreamService` orchestrates: **persist the user turn durable-first** → deterministic **`SafetyGate`** (on a separately-sanitized copy) → **classify** → route:

- **Green/Question** → `ComposeForConversationAsync` → `StreamAsync`, streamed as `token` frames, persisted as a coach turn on a clean finish with a terminal `done{turnId}`.
- **Red** (crisis/emergency) → scripted `988`/`741741` resources as a `safety` frame + coach turn; **no LLM** (classifier + stream never called).
- **Amber** (injury/RED-S) → scripted referral surfaced **before** the answer (distinct derived turn id, so it never collides with the answer turn), then the coach still answers.
- **WorkoutLog** → a `card` frame carrying the parsed draft + the server-resolved candidate prescription; **commits nothing** (the plan-mutating commit waits for an explicit Confirm — PR5).
- **Ambiguous** → a scripted clarification; never silently routed to a log.

Failure handling follows DEC-073/DEC-085: a classifier failure emits an `error` frame and persists nothing beyond the user turn (intent is never guessed); a mid/post-stream failure (`Transient`/`Permanent`/`Incomplete`) discards the partial, persists an **errored marker** (never a partial-as-complete), and emits an `error` frame; a **client abort** persists nothing and never 500s.

### Notable surface

- **`SseWriter`** — hand-rolled `event: …\ndata: {json}\n\n` framing + heartbeat, flush-per-frame, buffering off (no SSE dependency). The controller action returns `EmptyResult` after writing the stream manually.
- **`PostScriptedSafetyTurn`** command/handler — persists the Amber referral as a `CoachMessagePosted` coach turn (reusing the shipped event + projection; no `RegisteredEventTypes` change), keyed on a new `ConversationTurnId.DeriveSafetyTurnId` (distinct namespace from `DeriveCoachTurnId`).
- **`IWorkoutLogService.ResolveCandidatePrescriptionAsync`** — public seam over the existing prescription resolver for the card frame.
- **`IConversationFrame`** family (`token`/`safety`/`card`/`error`/`done`) — the orchestrator yields typed frames; the controller serializes.

> **Design note (worth a look):** the interactive stream (PR2) carries only User/Coach turns, so a chat safety turn is a **coach turn carrying scripted copy** (the scripted constants' own docs already call them "coach turns"). It renders as a coach bubble in the timeline at MVP-0; a tier-accented chat safety turn is deferred. Adding the `safety` SSE frame type + the `PostScriptedSafetyTurn` command is within PR4's mandate (wiring safety turns into the SSE surface), not in PR2's persistence schema.

### Tests (21 new, full suite 1979/1979 green in Replay)

- `SseWriterTests` (4) — framing, camelCase, heartbeat shape, per-frame flush.
- `ConversationTurnIdTests` (+5) — `DeriveSafetyTurnId` determinism + distinctness from the coach id.
- `ConversationMessagesEndpointIntegrationTests` (10) — Green, Green-no-plan, Red, Amber, WorkoutLog, Ambiguous, mid-stream failure, classifier failure, client abort, missing antiforgery.
- `StubCoachingLlm` gained a streaming test seam (`UseStreamBehavior`/`StreamCallCount`).

OpenAPI `swagger.json` + the frontend RTK/Zod codegen were regenerated for the new endpoint (drift gate clean).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01LCCmKo4SJT7isTLzyjwAEK

## Summary by CodeRabbit

* **New Features**
  * Added a new SSE-based endpoint for streaming conversation replies, including tokenized answers, safety notices, clarification prompts, workout confirmation cards, and terminal completion signaling.
  * Introduced a request DTO with required `clientMessageId` and message idempotency to prevent duplicate results on re-sends.
  * Documented the endpoint and DTO in the OpenAPI spec.
* **Bug Fixes**
  * Improved validation (antiforgery, authenticated user id, non-blank message, non-empty `clientMessageId`) and added structured retryable/non-retryable error frames for transient vs permanent failures.
* **Tests**
  * Added comprehensive end-to-end and SSE framing tests for the new streaming behavior.

### CODE REVIEW

File: backend/openapi/swagger.json [hunk 1]
```diff
@@ -473,6 +473,83 @@
...
+              "schema": {
+                "$ref": "#/components/schemas/ConversationMessageRequestDto"
+              }
+            }
+          }
+        },
+        "responses": {
+          "200": {
+            "description": "OK"
+          },
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/openapi/swagger.json:510]: Fixed in 6afba98 — added `[ProducesResponseType(typeof(string), StatusCodes.Status200OK, "text/event-stream")]` so the 200 advertises the SSE content block (string schema), while the 400/401 ProblemDetails responses keep their JSON content. (A blanket `[Produces("text/event-stream")]` was tried first but it mislabeled the error responses as SSE — the per-status overload is the correct tool.) `swagger.json` + the frontend RTK codegen are regenerated.

File: backend/openapi/swagger.json [hunk 2]
```diff
@@ -473,6 +473,83 @@
...
+                }
+              }
+            }
+          }
+        },
+        "security": [
+          {
+            "cookieAuth": [ ],
+            "bearerAuth": [ ]
+          }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/openapi/swagger.json:556]: Acknowledged, but this is a **pre-existing, repo-wide** pattern — not introduced by this PR. Every authenticated endpoint on `main` already emits `cookieAuth` + `bearerAuth` in one (ANDed) security object (verified against `/api/v1/auth/me`, `/api/v1/auth/logout`, `/api/v1/conversation/turns`). The new endpoint just follows the established convention. The correct fix is a central Swashbuckle `AddSecurityRequirement` change that splits the schemes into separate (OR) objects for **all** endpoints; scoping it to only this one endpoint would make the contract inconsistent. Declining here and tracking it as a separate cross-cutting follow-up.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs [hunk 3]
```diff
@@ -149,6 +156,85 @@ public async Task<IActionResult> GetTimeline(CancellationToken ct)
...
+                await writer.WriteEventAsync(
+                    "error",
+                    new ErrorFrame("Something went wrong on my end. Try again in a moment.", Retryable: true, RetryAfterSeconds: null),
+                    CancellationToken.None);
+            }
+            catch (Exception writeEx) when (writeEx is not OutOfMemoryException and not StackOverflowException)
+            {
+                // The stream may already be broken; nothing more can be done.
+                LogErrorFrameWriteFailed(logger, userId, writeEx);
+            }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs]: Fixed in 6afba98 — the catch now treats *any* exception as a clean disconnect once `RequestAborted` is signalled (`catch (Exception) when (ct.IsCancellationRequested)`), so an abort that surfaces as an `IOException` (broken pipe / connection reset) is handled the same as an `OperationCanceledException`. Genuine infrastructure failures (when the token is **not** cancelled) still fall through to the best-effort `error` frame + structured log.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationMessageRequestDto.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/// </summary>
+/// <param name="Message">The runner's free-text message. Sanitized server-side before it reaches the safety gate, the classifier, or the answer assembly.</param>
+/// <param name="ClientMessageId">
+/// Client-generated message id (typically <c>crypto.randomUUID()</c>). Marked
+/// <see cref="JsonRequiredAttribute"/> so System.Text.Json refuses to under-post a
+/// default <see cref="Guid"/> when the field is omitted.
+/// </param>
+public sealed record ConversationMessageRequestDto(
+    [property: JsonRequired] string Message,
+    [property: JsonRequired] Guid ClientMessageId);
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationMessageRequestDto.cs:21]: Fixed in 6afba98 — `PostMessage` now rejects a blank `Message` or an empty `ClientMessageId` with a structured 400 (ProblemDetails) **before** the stream opens. The empty-GUID guard matters because an empty client id would collide every empty-id post on the derived idempotency keys. Two integration tests cover it: `BlankMessage_IsRejectedWith400_BeforeAnyLlmCall` and `EmptyClientMessageId_IsRejectedWith400`.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationStreamService.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+    private async IAsyncEnumerable<IConversationFrame> StreamReplyCore(
+        Guid userId,
+        string message,
+        Guid clientMessageId,
+        [EnumeratorCancellation] CancellationToken ct)
+    {
+        // (1) Persist the user turn durably FIRST, before the gate or any stream, so a
+        //     crash after this point leaves a recoverable record. Idempotent on the client id.
+        await PersistUserTurnAsync(userId, clientMessageId, message, ct).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationStreamService.cs:66]: Deliberate MVP-0 tradeoff — declining for this slice with rationale, tracked as a follow-up.

The persistence is idempotent end-to-end (DEC-085 two-write): a same-id re-POST never corrupts state — both the durable-first user turn and the server-derived coach turn dedupe, so the stream holds exactly one of each (proven by `GreenQuestion_RePostedWithSameClientMessageId_IsIdempotent` and `AmberReferral_RePostedWithSameClientMessageId_DoesNotDoubleAppend`). The residual is a wasted model call + a transient *live-vs-persisted* text mismatch on a **same-id** double-submit — and D5 explicitly steers clients away from that (a re-send after a failure uses a **fresh** client id).

A full pre-stream replay path (detect the persisted coach turn, replay it as frames, skip the classifier/LLM) is a real heavy lift that also can't cleanly cover the errored-turn case, and for the self+family MVP-0 audience it isn't worth the added complexity. Captured as a deferred hardening item.

File: backend/tests/RunCoach.Api.Tests/Modules/Coaching/Conversation/Streaming/ConversationMessagesEndpointIntegrationTests.cs [hunk 6]
```diff
@@ -0,0 +1,788 @@
...
+        // Assert
+        var card = frames.Should().ContainSingle(f => f.Event == "card").Subject;
+        var payload = Payload<CardPayload>(card);
+        payload.Draft.DistanceValue.Should().Be(5);
+        payload.Draft.DistanceUnit.Should().Be(RunnerDistanceUnit.Kilometers);
+        frames.Should().NotContain(f => f.Event == "token", because: "a workout-log card streams no answer");
+
+        StubCoachingLlm.StreamCallCount.Should().Be(0, because: "the card path never streams");
+        (await CountWorkoutLogsAsync(userId)).Should().Be(0, because: "the card commits nothing until Confirm (PR5)");
+    }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/tests/RunCoach.Api.Tests/Modules/Coaching/Conversation/Streaming/ConversationMessagesEndpointIntegrationTests.cs:207]: Fixed in 6afba98 — the WorkoutLog test now asserts the `ConversationView` holds a single turn (the durable-first user turn, `Participant == User`), so a regression that appended a coach turn for the card path would fail here, not just the no-`WorkoutLog`-row check.

File: backend/tests/RunCoach.Api.Tests/Modules/Coaching/Conversation/Streaming/ConversationMessagesEndpointIntegrationTests.cs [hunk 7]
```diff
@@ -0,0 +1,788 @@
...
+            Content = JsonContent.Create(new ConversationMessageRequestDto("hi", Guid.NewGuid())),
+        };
+
+        // Act
+        using var response = await client.SendAsync(request, TestContext.Current.CancellationToken);
+
+        // Assert
+        response.StatusCode.Should().Be(
+            HttpStatusCode.BadRequest, because: "a mutation without the antiforgery token is rejected before streaming");
+    }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply backend/tests/RunCoach.Api.Tests/Modules/Coaching/Conversation/Streaming/ConversationMessagesEndpointIntegrationTests.cs:397]: Fixed in 6afba98 — the missing-antiforgery test now also asserts `application/problem+json`, and the two new boundary-validation tests assert the same structured ProblemDetails contract.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs [hunk 8]
```diff
@@ -149,6 +157,99 @@ public async Task<IActionResult> GetTimeline(CancellationToken ct)
...
+            // tokens then flow continuously and keep the connection alive themselves.
+            await writer.WriteHeartbeatAsync(ct);
+            await foreach (var frame in streamService
+                .StreamReplyAsync(userId, request.Message, request.ClientMessageId, ct)
+                .WithCancellation(ct))
+            {
+                await writer.WriteEventAsync(frame.EventName, frame, ct);
+            }
+        }
+        catch (Exception) when (ct.IsCancellationRequested)
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs]: **💡 [LOW] Over-broad cancellation catch swallows genuine server bugs on client disconnect**

The first catch in `PostMessage` is `catch (Exception) when (ct.IsCancellationRequested)` with an empty body. The `when` filter is unqualified by exception **type**, so any exception thrown from the orchestrator while the request token is already cancelled is swallowed with zero observability — a real bug (NRE, `InvalidCastException`) during a cancelled request becomes indistinguishable from a clean disconnect, while the generic catch below it logs via `LogUnhandledStreamError`. For a chat endpoint where runners frequently navigate away mid-stream, the cancelled-token window is common. Timing-dependent, hence low severity. Narrow the filter to the expected disconnect types so genuine bugs still fall through to the logging catch:

```suggestion
        catch (Exception ex) when (ct.IsCancellationRequested && ex is OperationCanceledException or IOException)
```

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/DoneFrame.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+namespace RunCoach.Api.Modules.Coaching.Conversation.Streaming;
+
+/// <summary>
+/// The terminal success frame (Slice 4B PR4) emitted as <c>event: done</c> after a coach
+/// turn (answer, scripted safety turn, clarification, or confirmation card) completes.
+/// Carries the persisted <see cref="TurnId"/> — load-bearing for the client's
+/// reconcile-once: the live bubble is folded into the timeline cache keyed on this id.
+/// </summary>
+/// <param name="TurnId">The persisted coach/safety/card turn id (server-derived for the coach reply).</param>
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/DoneFrame.cs]: **💡 [LOW] DoneFrame doc claims a done frame follows a confirmation card, but the card path emits none**

The XML doc lists "confirmation card" among turns that complete with a `done` frame and documents `TurnId` as "the persisted coach/safety/card turn id". This is inaccurate: the `MessageIntent.WorkoutLog` branch in `ConversationStreamService` yields a `CardFrame` then `yield break` with **no** `DoneFrame`, and the card commits nothing (no persisted turn id exists). The `CardFrame` doc and the `WorkoutLog_ReturnsConfirmationCard_AndCommitsNothing` test both confirm this.

**Suggested fix:** Remove "confirmation card" from the summary and drop "card" from the `TurnId` param description — the card path emits no done frame and persists no turn id.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/CandidatePrescriptionDto.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    public static CandidatePrescriptionDto? FromSnapshot(WorkoutPrescriptionSnapshot? snapshot) =>
+        snapshot is null
+            ? null
+            : new CandidatePrescriptionDto(
+                snapshot.WorkoutType.ToString(),
+                snapshot.PrescribedDistance.Meters,
+                snapshot.PrescribedDuration.TotalSeconds,
+                snapshot.PrescribedPaceFast.SecondsPerKm,
+                snapshot.PrescribedPaceSlow.SecondsPerKm);
+}
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/CandidatePrescriptionDto.cs:37]: **💡 [LOW] Card Prescription projection (CandidatePrescriptionDto.FromSnapshot) is never asserted**

The WorkoutLog card's server-resolved candidate prescription is never asserted. `WorkoutLog_ReturnsConfirmationCard_AndCommitsNothing` seeds an active plan and asserts only `card.Draft.DistanceValue`/`DistanceUnit`, never `card.Prescription`. `FromSnapshot` maps five fields (incl. `PrescribedPaceFast`→`PaceFastSecPerKm` and `PrescribedPaceSlow`→`PaceEasySecPerKm`) and returns `null` off-plan; neither the populated projection nor the null branch is exercised, and there is no `CandidatePrescriptionDtoTests`. A swapped pace-band mapping would ship undetected.

**Suggested fix:** In the WorkoutLog test, seed a plan slot matching the logged date and assert `card.Prescription` is non-null with the expected fields and `fast < easy` pace ordering; add a second off-plan scenario asserting `null`. Optionally add a `CandidatePrescriptionDtoTests` unit covering `FromSnapshot` and the null-snapshot branch.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs [hunk 11]
```diff
@@ -149,6 +157,99 @@ public async Task<IActionResult> GetTimeline(CancellationToken ct)
...
+                    "error",
+                    new ErrorFrame("Something went wrong on my end. Try again in a moment.", Retryable: true, RetryAfterSeconds: null),
+                    CancellationToken.None);
+            }
+            catch (Exception writeEx) when (writeEx is not OutOfMemoryException and not StackOverflowException)
+            {
+                // The stream may already be broken; nothing more can be done.
+                LogErrorFrameWriteFailed(logger, userId, writeEx);
+            }
+        }
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs:250]: **💡 [LOW] Controller post-heartbeat unhandled-error path (best-effort error frame) is untested**

The `catch (Exception ex)` after the heartbeat commits the 200 logs and writes a best-effort `error` frame when an infrastructure failure (bus / Marten / EF) escapes `StreamReplyAsync` — e.g. `PersistUserTurnAsync` throwing. But every failure scenario in the suite is scripted inside the LLM stream (`Transient`/`Permanent`/`Incomplete` `CoachingLlmException`), all caught **inside** `ConversationStreamService`, never reaching this controller catch. Nothing forces the user-turn persistence or context loader to throw, so this error-frame emission and the nested write-failure handler are uncovered.

**Suggested fix:** Add a test that forces an orchestrator infra failure after the heartbeat (e.g. a throwing `IConversationStreamService`/`IMessageBus` test double) and assert the response still emits an `error` frame (no torn connection, no 500) and that `LogUnhandledStreamError` fires.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationContextLoader.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        var conversation = await session.LoadAsync<ConversationView>(userId, ct).ConfigureAwait(false);
+        IReadOnlyList<ConversationContextTurn> recentTurns = (conversation?.Turns ?? [])
+            .Where(t => !t.IsErrored && t.TurnId != clientMessageId)
+            .TakeLast(RecentTurnLimit)
+            .Select(t => new ConversationContextTurn(t.Participant, t.Content))
+            .ToArray();
+
+        var logEntities = await workoutLogs
+            .GetByUserAsync(userId, cursor: null, limit: RecentLogLimit, ct)
+            .ConfigureAwait(false);
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationContextLoader.cs:66]: **💡 [LOW] Recent-turns context filtering (exclude errored markers and current turn) is untested**

`LoadAnswerContextAsync` builds the LLM grounding with `Where(t => !t.IsErrored && t.TurnId != clientMessageId).TakeLast(10)` (plus a 5-log limit), excluding errored markers and the just-persisted current user turn. Because `StubCoachingLlm.StreamAsync` ignores the prompt, no test inspects the composed context, so this filtering is unverified — a regression feeding errored markers or the duplicated current turn into the prompt would pass every test.

**Suggested fix:** Add a focused `ConversationContextLoader` test (or a recording `IContextAssembler` double) that persists an errored coach turn plus prior turns and verifies the loaded `RecentTurns` excludes the errored marker and the current user turn, capped at the limit.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/SseWriter.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        ArgumentNullException.ThrowIfNull(data);
+
+        // Compact (no indentation) so the JSON occupies a single `data:` line — a frame
+        // is terminated by the blank line, so an embedded newline would split the frame.
+        var json = JsonSerializer.Serialize(data, _jsonOptions);
+        var frame = Encoding.UTF8.GetBytes($"event: {eventName}\ndata: {json}\n\n");
+
+        await _output.WriteAsync(frame, ct).ConfigureAwait(false);
+        await _output.FlushAsync(ct).ConfigureAwait(false);
+    }
```
Comment (leehopper) [inline backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/SseWriter.cs:43]: **💡 [LOW] SseWriter.WriteEventAsync re-decouples the SSE event name from the frame (contested)**

**Contested / optional (confidence 15%).** `IConversationFrame.EventName` is an explicit-interface member designed to bind the SSE event name to each frame so name and payload can't diverge. `WriteEventAsync(string eventName, object data, ct)` accepts the name and payload independently (payload typed `object`), so nothing at compile time stops a caller emitting an event name that contradicts the payload (e.g. `event: token` carrying a `DoneFrame`). Both call sites are consistent today; the validator considered `SseWriter` a defensible generic SSE primitive, the challenger found the observation factually accurate.

**Optional refactor:** `WriteEventAsync(IConversationFrame frame, CancellationToken ct)` reading `frame.EventName` internally would restore the invariant (keep `WriteHeartbeatAsync` separate). Skip if you intend `SseWriter` to stay frame-agnostic.

**Review by leehopper — COMMENTED** [review]
### Deep Review

Reviewed **PR #233** (Slice 4B SSE conversation endpoint) with a 7-agent Frontier (Opus) pipeline + Sonnet validation + Opus blind challenge. This is a careful, well-tested PR — **security, cross-file impact, and simplification came back clean, nothing blocking**.

Found **6 low-severity items**: 1 correctness/observability nit + 5 improvement suggestions (3 test-coverage gaps, 1 doc inaccuracy, 1 contested type-design note). Each is posted as an inline comment below.

| Finding | Conf | What |
|---|---|---|
| 💡 bug-1 | 70% | Over-broad `catch (Exception) when (ct.IsCancellationRequested)` swallows genuine bugs on disconnect |
| 💡 conv-1 | 70% | `DoneFrame` doc claims the confirmation-card path emits a done frame; it doesn't |
| 💡 test-1 | 65% | Card `Prescription` projection (`FromSnapshot`) never asserted |
| 💡 test-2 | 65% | Controller post-heartbeat error-frame path untested |
| 💡 test-3 | 57% | Recent-turns context filtering (errored/current-turn exclusion) untested |
| 💡 type-1 | 15% | `SseWriter.WriteEventAsync` decouples event name from frame (contested) |

---
Generated by deep-review | Reviewed up to: 6afba981a06cd0b471c787870ae4fa5d25fcbbde

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs (hunk 8 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs]: Fixed in 7049513 — narrowed to `catch (Exception ex) when (ct.IsCancellationRequested && ex is OperationCanceledException or IOException)`. Those two are exactly the documented disconnect signals (an `OperationCanceledException`, or an `IOException` / `ConnectionResetException` for a broken pipe or reset), so a client abort is still handled cleanly, while a genuine bug thrown during a cancelled request (an NRE, an `InvalidCastException`) now falls through to the logging `catch (Exception ex)` below instead of being swallowed. The new `PostHeartbeatUnhandledError_...` test exercises that fall-through path.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/DoneFrame.cs (hunk 9 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/DoneFrame.cs]: Fixed in 7049513 — removed "confirmation card" from the summary and dropped "card" from the `TurnId` param doc. The `MessageIntent.WorkoutLog` branch yields a `CardFrame` then `yield break` (no `DoneFrame`) and persists no coach turn, so the card path has no done frame and no persisted turn id; the doc now states that explicitly.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/CandidatePrescriptionDto.cs (hunk 10 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/CandidatePrescriptionDto.cs:37]: Added `CandidatePrescriptionDtoTests` in 7049513. It covers `FromSnapshot` with **distinct** fast/easy bounds (fast = 240, easy = 330 s/km) and asserts each field maps correctly — in particular `PaceFastSecPerKm` ← `PrescribedPaceFast`, `PaceEasySecPerKm` ← `PrescribedPaceSlow`, and `fast < easy` — so a swapped pace-band projection fails loudly; a second case covers the off-plan `null` branch. That is the robust guard against the swap this finding is about.

I also tried the suggested end-to-end `card.Prescription` assertion, and it surfaced a separate **test-harness** issue (not a product bug): `SeedActivePlanAsync` seeds `RunnerOnboardingProfile` with an explicit `TenantId = userId.ToString()`, and that entity is `ITenanted` behind a Wolverine EF tenant filter, so the **authenticated SSE request** filters the row out (`CurrentPlanId` → null → "No active plan" → null prescription). `ConversationTimelineControllerIntegrationTests` seeds the profile **without** an explicit `TenantId` and its request reads the plan fine. The other SSE "seed active plan" tests don't notice because the answer-grounding stub ignores the prompt. In production the profile is created under the request's own tenant, so this only affects the test seam. Aligning `SeedActivePlanAsync` with the timeline test's seeding and restoring the end-to-end card-prescription assertion is a small follow-up; the unit test fully guards the mapping in the meantime.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs (hunk 11 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/ConversationController.cs:250]: Added `PostHeartbeatUnhandledError_EmitsBestEffortRetryableErrorFrame_FromTheControllerCatch` in 7049513. It scripts the classifier to throw a non-`CoachingLlmException` (an `InvalidOperationException` standing in for a bus / Marten / EF fault): the orchestrator's in-service catch only handles `CoachingLlmException`, so it escapes `StreamReplyAsync` into the controller's post-heartbeat `catch (Exception)`. The test asserts 200 + `text/event-stream`, a single clean `error` frame (`Retryable: true`, `RetryAfterSeconds: null`), no `done` / `token`, the controller's exact best-effort copy (proving it's the controller catch, not the in-service `ClassifierFailedMessage` path), and that only the durable user turn persisted. It reuses the existing static `StubCoachingLlm` seam, so no second host is booted.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationContextLoader.cs (hunk 12 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/ConversationContextLoader.cs:66]: Added `LoadAnswerContext_ExcludesErroredAndCurrentTurns_AndCapsAtTheRecentLimit` in 7049513. It seeds a `ConversationView` through the same `PostUserConversationTurn` / `PostCoachConversationTurn` bus commands production uses (so the real `InteractiveConversationProjection` materializes it): 12 ordinary turns, one errored coach marker, and the current user turn keyed by the `clientMessageId` under test. It then resolves `IConversationContextLoader` and asserts `RecentTurns` excludes the errored marker (`!IsErrored`) and the current turn (`TurnId != clientMessageId`), and keeps only the most-recent 10 (turn-02..turn-11) in append order — so a regression that fed errored markers or the duplicated current turn into the LLM grounding would fail.

File: backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/SseWriter.cs (hunk 13 above)
Comment (leehopper) [reply backend/src/RunCoach.Api/Modules/Coaching/Conversation/Streaming/SseWriter.cs:43]: Declining this one — it was flagged contested / optional (~15%), and keeping `SseWriter` frame-agnostic is the better call. The name↔payload binding is already enforced at the frame-definition site: `IConversationFrame.EventName` is an explicit-interface member, so a new frame type can't be added without supplying its event name, and both call sites are consistent today. A frame-typed `WriteEventAsync(IConversationFrame, ct)` overload would also introduce a System.Text.Json sharp edge — the current `object data` parameter serializes the **runtime** type, but an `IConversationFrame`-typed parameter serializes the **declared** interface (whose only member is the ignored explicit `EventName`), yielding empty `{}` payloads unless the runtime type is passed explicitly. Net: low value, real coupling, plus a footgun — so I'm leaving `SseWriter` as a generic SSE wire primitive.
